src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -243,6 +243,35 @@ struct md_mmap_info {
 	size_t  size;
 };
 
+struct md_apk_mmap_info {
+	md_mmap_info mmap_info;
+	md_mmap_info file_info;
+};
+
+MONO_API int monodroid_getpagesize (void);
+
+static md_apk_mmap_info
+md_mmap_apk_file(int fd, uLong offset, uLong size, const char* filename, const char* apk)
+{
+	md_apk_mmap_info info;
+	
+	int pageSize = monodroid_getpagesize();
+	uLong offsetFromPage = offset % pageSize;
+	uLong offsetPage = offset - offsetFromPage;
+	uLong offsetSize = size + offsetFromPage;
+	
+	info.mmap_info.area = mmap(NULL, offsetSize, PROT_READ, MAP_PRIVATE, fd, offsetPage);
```

Comment:
There should be some form of error detection around this; **mmap**(2) *can* fail, hence issue #1673.  That said, the `mmap()` is *required* in order for anything to work, so we should check for the error and fail with a more appropriate error message than what we currently do: crash with a SIGSEGV.

```suggestion
	info.mmap_info.area = mmap (NULL, offsetSize, PROT_READ, MAP_PRIVATE, fd, offsetPage);
	if (info.mmap_info.area == MAP_FAILED) {
	    log_fatal (LOG_DEFAULT, "Could not `mmap` apk `%s` entry `%s`: %s", app, filename, strerror (errno));
	    exit (FATAL_EXIT_CANNOT_FIND_APK);
	}
```

Reply:
Indeed, good catch. Implemented and tested in 4a08ef9